### PR TITLE
icn2img tool: handle multiple ICN files

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
Related to #6553

On Windows, only globs in the filename part of a path are supported (`dir/arch*.?cn`); globs in the intermediate directory names (`*/filename.icn`) are not supported. On *NIX there are no such limitations.

### CODE REVIEW

File: src/tools/icn2img.cpp [hunk 1]
```diff
@@ -21,103 +21,207 @@
...
+                dstFileName += ".bmp";
             }
 
-            std::cout << "Image " << ii + 1 << " has offset of [" << static_cast<int32_t>( head.offsetX ) << ", " << static_cast<int32_t>( head.offsetY ) << "]"
-                      << std::endl;
+            static_assert( std::is_same_v<decltype( header.offsetX ), uint16_t> && std::is_same_v<decltype( header.offsetY ), uint16_t>,
+                           "Offset types have been changed, check the casts below" );
 
-            fheroes2::Save( image, dstfile, 23 );
+            std::cout << "Image " << spriteIdx + 1 << " has offset of [" << static_cast<int16_t>( header.offsetX ) << ", " << static_cast<int16_t>( header.offsetY )
```
Comment (oleg-derevenetz) [inline src/tools/icn2img.cpp:219]: These offsets are in fact signed, right?

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , as a separate solution I am using a batch file on Windows🙂

**Comment by oleg-derevenetz** [comment]
> as a separate solution I am using a batch file on Windows🙂

This is slow, because a separate instance of `icn2img` is running for each ICN file. Also on *NIX globs will not work either, because, although on *NIX shell does all the hard work with globs, this still requires the application to be able to accept multiple arguments with file names, which `icn2img` is not able to do in its current form.

**Comment by oleg-derevenetz** [comment]
@ihhub BTW is this normal that extracted PNG images are absolutely transparent? They looks fine in Explorer preview (it seems that it doesn't handle transparency very well), but completely transparent in most viewers.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , performance is not an issue here as this tool is not a part of the main project. It is good to have but it is optional.

Some PNG files are totally transparent if they are 1x1 pixel size but 99% of them have some visible information.

**Comment by ihhub** [comment]
@oleg-derevenetz , I use **XnView** and even normal Window Viewer works fine to me.

File: src/tools/icn2img.cpp (hunk 1 above)
Comment (ihhub) [reply src/tools/icn2img.cpp:217]: Actually, yes. We have to fix it.

**Comment by oleg-derevenetz** [comment]
@ihhub

> I use XnView and even normal Window Viewer works fine to me.

I have a weird issue with PNG images (BMP images are fine). PNG files are of the right size, but entirely transparent. What if you open PNG file in Paint for example? I tried the files from `archer.icn`.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , I compiled the latest version of these tools from master branch for SDL 2 and everything works. I can see PNG file via XnView. This is a batch file I used to extract all images (yeah, it's an absolute mess but it works):
```bat
@echo off

for /R %%f in (*.AGG) do (
 mkdir "%%~nf"
 extractor.exe "%%~nf.AGG" "%%~nf"
 
 for /R %%i in (%%~nf\*.icn) do (
  mkdir "%%~nf\%%~ni"
  icn2img.exe "%%~nf\%%~ni.icn" . "%%~nf\KB.PAL"
 )
)
```

**Comment by ihhub** [comment]
Paint shows proper images as well:
image

**Comment by oleg-derevenetz** [comment]
@ihhub

>  I compiled the latest version of these tools from master branch for SDL 2 and everything works.

Well, in my case extracted PNG files look strange. In Windows Explorer:

explorer

But for example in MS Paint they are completely transparent :) I attached the archive with PNG images. Are they OK on your system?

[archer.zip](https://github.com/ihhub/fheroes2/files/10611417/archer.zip)

**Comment by ihhub** [comment]
Your images are very weird:
image

I also noticed that your images are bigger. Do you use SDL 2 tools? SDL 1 doesn't support PNG and these images by size look like Bitmap.

**Comment by ihhub** [comment]
Also please try a full rebuild of the tool. Just in case...

**Comment by oleg-derevenetz** [comment]
@ihhub

> Do you use SDL 2 tools?

Yes, and these files are really PNG - at least they have proper PNG header.

**Comment by oleg-derevenetz** [comment]
And here are BMP files (just replaced ".png" with ".bmp" in `icn2img.cpp`):

[archer.zip](https://github.com/ihhub/fheroes2/files/10611480/archer.zip)

It seems that they have no issues.

**Comment by oleg-derevenetz** [comment]
@ihhub Huh, I just removed the `libpng16.dll` from build dir and now these PNGs are fine. It seems that SDL tries to use `libpng` if present (it seems that otherwise it uses some internal PNG library implementation) and this somehow messes things up.

**Comment by ihhub** [comment]
Wow! This is really bad. That is why I created an issue to remove dependencies from SDL about loading and saving images but this is another story. We can not to include unused libraries into the final package for tools.

**Comment by ihhub** [comment]
Let's remove this library to avoid messing things.

**Comment by oleg-derevenetz** [comment]
> Let's remove this library to avoid messing things.

I patched `vcpkg` port to force the `sdl2-image` to use the built-in PNG library and removed the `libpng` completely. I tested `icn2img` and can confirm that issue is gone.

**Review by ihhub — APPROVED** [review]

**Pull request merged**